### PR TITLE
JIT: rework phase objects so we can use them more widely

### PR DESCRIPTION
**Pull request by AndyAyersMS**
Next part of #2109.

Add support for phases that are simply compiler methods or lambdas. These are
not used yet -- the plan is to introduce new phases gradually, cleaning up
redundant checking and dumping along the way. This will happen in subequent
changes.

Remove a bit of now-redundant post-phase dumping and checking from lower.

Add the active phase to the assertion text, so we have some context.

### CODE REVIEW

**Comment by AndyAyersMS** [comment]
cc @dotnet/jit-contrib 

No diffs; jit dump is a bit smaller as there were basically two IR dumps at the end of lower.

**Review by CarolEidt — APPROVED** [review]
LGTM - I really like the direction this is going! I'm not sure I would have added the method/lambda version until it's used, but no harm I guess.

**Comment by AndyAyersMS** [comment]
I have another change that phasifies everything (and uses the new phase forms introduced here) but it currently introduces a lot of redundant dumping and checking.

I do not want to make checked builds slower or use more memory or make jit dump files larger. So I have to work towards this bit by bit.

The next few steps are:
* introduce check auditing so we know what post-phase checks we run currently
* introduce some simple ways of indicating that a phase did not modify anything and suppress redundant post-phase dumps and checks. Initially this will just catch the trivial cases where the phase itself is certain it did not make changes. Later we can consider trying to auto-detect this somehow.
* gradually start introducing phase objects to wrap things that should be phases. Verify via check auditing that the jit is doing the same checks it does now. Remove redundant manually invoked checks or dumps.
* hopefully enough of the new phase-ending checks and dumps are redundant and we can do the above without impacting TP/mem/dump file size.
* break checks into finer granularity parcels and enable the simpler checks as early as possible.

**Comment by BruceForstall** [comment]
Maybe you could start by having a phase opt-in to post-phase checks/dumps that it already does, remove those manual steps, and have the phase infrastructure do them.

**Comment by AndyAyersMS** [comment]
I had originally thought to strongly couple phases to post-phase checks, but didn't like how this came out, it was a lot of extra detail. Current proposal is that the set of active post-phase checks will be part of the ambient compiler state, so as we progress through the phase list we will add / remove active checks. We can do this either in between or during phases.

On top of this, I may add a way for a phase to request "additional" checks, if there are phases that warrant extra scrutiny and the extra checks are not specific to just one phase.

File: src/coreclr/src/jit/CMakeLists.txt [hunk 1]
```diff
@@ -360,31 +360,23 @@ function(add_jit jitName)
...
-
     target_link_libraries(${jitName}
         ${JIT_LINK_LIBRARIES}
         ${JIT_ARCH_LINK_LIBRARIES}
     )
-
     # add the install targets
     install_clr(TARGETS ${jitName})
 endfunction()
-
```
Comment (BruceForstall) [inline src/coreclr/src/jit/CMakeLists.txt]: hmmm... you don't like whitespace?

**Review by BruceForstall — APPROVED** [review]
LGTM

File: src/coreclr/src/jit/CMakeLists.txt (hunk 1 above)
Comment (AndyAyersMS) [reply src/coreclr/src/jit/CMakeLists.txt]: Didn't mean to do that. Let me undo that part.

**Pull request merged**